File: 1/2020.10.26/547.Friend Circles/python/solution.py
class Solution:
    def findCircleNum(self, M: List[List[int]]) -> int:
        # DFS
        def DFS(i):
            for j in range(len(M)):
                if M[i][j] == 1 and not visited[j]:
                    visited[i] = 1
                    DFS(j)
        visited = [0 for _ in range(len(M))]
        count = 0
        for i in range(len(M)):
            if not visited[i]:
                visited[i] = 1
                DFS(i)
                count += 1
        return count

    def findCircleNum(self, M: List[List[int]]) -> int:
        # BFS
        visited = [0 for _ in range(len(M))]
        count, q = 0, []
        for i in range(len(M)):
            if not visited[i]:
                q.append(i)
                while q:
                    s = q.pop(0)
                    visited[s] = 1
                    for j in range(len(M)):
                        if M[s][j] == 1 and not visited[j]:
                            visited[j] = 1
                            q.append(j)
                count += 1
        return count

File: 1/2020.10.26/547.Friend Circles/java/solution.java
class Solution {
    public int findCircleNum(int[][] M) {
        // DFS
        int[] visited = new int[M.length];
        int count = 0;
        for (int i = 0; i < M.length; i++) {
            if (visited[i] == 0) {
                visited[i] = 1;
                DFS(M, visited, i);
                count++;
            }
        }
        return count;
    }
    
    private void DFS(int[][] m, int[] visited, int i) {
        for (int j = 0; j < m.length; j++) {
            if (m[i][j] == 1 && visited[j] == 0) {
                visited[j] = 1;
                DFS(m, visited, j);
            }
        }
    }

    public int findCircleNum(int[][] M) {
        // BFS
        int[] visited = new int[M.length];
        int count = 0;
        Deque<Integer> q = new ArrayDeque<>();
        for (int i = 0; i < M.length; i++) {
            if (visited[i] == 0) {
                q.offerLast(i);
                while (!q.isEmpty()) {
                    int s = q.pollFirst();
                    visited[s] = 1;
                    for (int j = 0; j < M.length; j++) {
                        if (M[s][j] == 1 && visited[j] == 0) {
                            visited[j] = 1;
                            q.offerLast(j);
                        }
                    }
                }
                count++;
            }
        }
        return count;
    }
}
